### PR TITLE
docs: Improve CPU prover setup documentation

### PR DESCRIPTION
**Pull request by jrchatruc**
# What ❔

This PR expands the documentation on how to run the stack with the CPU prover, also adding a small script that will automatically generate the setup data needed so it's more streamlined.

## Checklist

- [x] PR title corresponds to the body of PR (we generate changelog entries from PRs).
- [ ] Tests for the changes have been added / updated.
- [x] Documentation comments have been added / updated.
- [x] Code has been formatted via `zk fmt` and `zk lint`.

### CODE REVIEW

File: prover/setup.sh [hunk 1]
```diff
@@ -0,0 +1,20 @@
+#!/usr/bin/env bash
+cd ..
+export ZKSYNC_HOME=$(pwd)
```
Comment (mm-zk) [inline prover/setup.sh:3]: setting this environment variable here is 'surprising' - let's check its value instead and complain if this is set incorrectly.

File: prover/setup.sh [hunk 2]
```diff
@@ -0,0 +1,20 @@
+#!/usr/bin/env bash
```
Comment (mm-zk) [inline prover/setup.sh:1]: add a small comment on what this script does (that it sets up necessary data for the CPU fri prover locally).

File: prover/setup.sh [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+cd ..
+export ZKSYNC_HOME=$(pwd)
+cd prover
+
+sed -i '' 's/^ETH_SENDER_SENDER_PROOF_SENDING_MODE.*$/ETH_SENDER_SENDER_PROOF_SENDING_MODE=OnlyRealProofs/' ../etc/env/dev.env
```
Comment (mm-zk) [inline prover/setup.sh:6]: dev.env file can be overwritten by the scripts.

the 'zk config compile dev' is taking data from dev.toml (and other places) and writes to dev.env

so I'd suggest maybe updating dev.toml, and then calling `zk config compile dev` ?

File: prover/prover_fri/README.md [hunk 4]
```diff
@@ -17,34 +17,46 @@ Machine specs:
 - RAM: 60GB of RAM(if you have lower RAM machine enable swap)
 - Disk: 400GB of free disk
 
+Before starting, make sure you go into the root of the repository, then run
+
+```
+export ZKSYNC_HOME=$(pwd)
```
Comment (mm-zk) [inline prover/prover_fri/README.md:23]: after this PR gets in: https://github.com/matter-labs/zksync-era/pull/188 we should have more clear error for users.

File: prover/setup.sh [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+sed -i '' 's/^ETH_SENDER_SENDER_PROOF_LOADING_MODE.*$/ETH_SENDER_SENDER_PROOF_LOADING_MODE=FriProofFromGcs/' ../etc/env/dev.env
+sed -i '' 's/^FRI_PROVER_SETUP_DATA_PATH.*$/FRI_PROVER_SETUP_DATA_PATH=vk_setup_data_generator_server_fri\/data\//' ../etc/env/dev.env
+sed -i '' 's/^FRI_PROOF_COMPRESSOR_UNIVERSAL_SETUP_PATH.*$/FRI_PROOF_COMPRESSOR_UNIVERSAL_SETUP_PATH=..\/keys\/setup\/setup_2^26.key/' ../etc/env/dev.env
+sed -i '' 's/^FRI_PROOF_COMPRESSOR_PROMETHEUS_PUSH_INTERVAL_MS.*$/FRI_PROOF_COMPRESSOR_PROMETHEUS_PUSH_INTERVAL_MS=100000000/' ../etc/env/dev.env
```
Comment (mm-zk) [inline prover/setup.sh:10]: we have fixed the underlying library that was spamming the logs.
So when you sync your boojum-integration branch past https://github.com/matter-labs/zksync-era/pull/253 - the logs should not show up anymore

**Review by mm-zk — COMMENTED** [review]
we are currently in the process of merging the boojum-branch into the main.

And the 'prover' directory is fully merged, that's why I'd suggest changing this PR to add these into 'main' branch first (and then we'll bring them over to boojum-branch when we sync)

File: prover/setup.sh (hunk 1 above)
Comment (jrchatruc) [reply prover/setup.sh]: Fixed here [e7b613f](https://github.com/matter-labs/zksync-era/pull/258/commits/e7b613f39fd439668bada0b8ee988970b39bef54)

File: prover/setup.sh (hunk 2 above)
Comment (jrchatruc) [reply prover/setup.sh:1]: Added here [1f3dd01](https://github.com/matter-labs/zksync-era/pull/258/commits/1f3dd014e0db01124e80d1c2e4c4ad1d148a63d4)

File: prover/setup.sh (hunk 3 above)
Comment (jrchatruc) [reply prover/setup.sh]: Makes sense, i changed it here [9ea0427](https://github.com/matter-labs/zksync-era/pull/258/commits/9ea0427803f4f18e317e61ede60b008d68c65c1d)

File: prover/setup.sh (hunk 5 above)
Comment (jrchatruc) [reply prover/setup.sh]: Cool, i remove the prometheus push internal change here [478d130](https://github.com/matter-labs/zksync-era/pull/258/commits/478d130dec5f1707cf71c9c82f46a8db2bd67393)

File: prover/prover_fri/README.md (hunk 4 above)
Comment (jrchatruc) [reply prover/prover_fri/README.md:23]: nice, I think it's still ok to leave the comment though

File: prover/setup.sh [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+sed -i.backup 's/^setup_data_path=.*$/setup_data_path="vk_setup_data_generator_server_fri\/data\/"/' ../etc/env/base/fri_prover.toml
+rm ../etc/env/base/fri_prover.toml.backup
+sed -i.backup 's/^universal_setup_path=.*$/universal_setup_path="..\/keys\/setup\/setup_2^26.key"/' ../etc/env/base/fri_proof_compressor.toml
+rm ../etc/env/base/fri_proof_compressor.toml.backup
+
+zk config compile dev
+
+for i in {1..13}
+do
+    zk f cargo run --release --bin zksync_setup_data_generator_fri -- --numeric-circuit $i --is_base_layer
```
Comment (mm-zk) [inline prover/setup.sh:21]: these take a lot of time -- I wonder if we should somehow detect that the files are already present?

File: prover/setup_gpu.sh [hunk 7]
```diff
@@ -0,0 +1,27 @@
+#!/usr/bin/env bash
+# This script sets up the necessary data needed by the GPU FRI prover to be used locally.
+
+if [[ -z "${ZKSYNC_HOME}" ]]; then
+  echo "Environment variable ZKSYNC_HOME is not set. Make sure it's set and pointing to the root of this repository"
+  exit 1
+fi
+
+sed -i.backup 's/^proof_sending_mode=.*$/proof_sending_mode="OnlyRealProofs"/' ../etc/env/base/eth_sender.toml
```
Comment (mm-zk) [inline prover/setup_gpu.sh:9]: lots of stuff is common with setup.sh .. maybe we should have a common script (with a flag), or 2 files, but they call common utils?

File: prover/witness_generator/src/main.rs [hunk 8]
```diff
@@ -39,7 +39,10 @@ struct Opt {
     batch_size: Option<usize>,
     /// aggregation round for the witness generator.
     #[structopt(short = "r", long = "round")]
-    round: AggregationRound,
+    round: Option<AggregationRound>,
+    /// start all aggregation rounds for the witness generator.
```
Comment (mm-zk) [inline prover/witness_generator/src/main.rs:43]: please add more descriptive comments -- mention that the binary can either be run for a single aggregation round, or for all rounds.

**Comment by mm-zk** [comment]
LGTM - pinging @akash-chandrakar to PTAL on the changes to the witness binary.

File: prover/setup.sh (hunk 6 above)
Comment (Juan-M-V) [reply prover/setup.sh]: I added some basic detection for this in the setup script. https://github.com/matter-labs/zksync-era/pull/258/commits/3435d1b8ce22056cad473965b937e3e7f59a1c72

File: prover/setup_gpu.sh (hunk 7 above)
Comment (Juan-M-V) [reply prover/setup_gpu.sh]: I added a flag to the script to be able to pick between cpu and gpu. https://github.com/matter-labs/zksync-era/pull/258/commits/3435d1b8ce22056cad473965b937e3e7f59a1c72 & https://github.com/matter-labs/zksync-era/pull/258/commits/54b2c0ebb11d6617a46de18017887b1079c41057

File: prover/witness_generator/src/main.rs (hunk 8 above)
Comment (Juan-M-V) [reply prover/witness_generator/src/main.rs]: I added this line to the Opt struct: `Aggregation rounds options, they can be run individually or together.` https://github.com/matter-labs/zksync-era/pull/258/commits/3435d1b8ce22056cad473965b937e3e7f59a1c72

File: prover/witness_generator/src/main.rs [hunk 9]
```diff
@@ -124,85 +120,112 @@ async fn main() -> anyhow::Result<()> {
...
+        (Some(round), false) => vec![round],
+        (None, true) => vec![
+            AggregationRound::BasicCircuits,
+            AggregationRound::LeafAggregation,
+            AggregationRound::NodeAggregation,
+            AggregationRound::Scheduler,
+        ],
+        _ => {
+            return Err(anyhow!(
+                "Expected --all_rounds flag with no --round flag present"
```
Comment (mm-zk) [inline prover/witness_generator/src/main.rs:133]: nit: this is not fully correct - as this condition also triggers, when you specify both --all_rounds and --round=X flag.

File: prover/setup.sh [hunk 10]
```diff
@@ -0,0 +1,44 @@
+#!/usr/bin/env bash
+# This script sets up the necessary data needed by the CPU FRI prover to be used locally.
```
Comment (mm-zk) [inline prover/setup.sh:2]: nit - seems you both support CPU and GPU

**Review by mm-zk — DISMISSED** [review]

File: prover/setup.sh [hunk 11]
```diff
@@ -0,0 +1,44 @@
+#!/usr/bin/env bash
+# This script sets up the necessary data needed by the CPU/GPU FRI prover to be used locally.
+
+GPU_FLAG=""
+if [ "$1" = "gpu" ]; then
+    GPU_FLAG='--features "gpu"'
```
Comment (mm-zk) [inline prover/setup.sh:6]: BTW - error: none of the selected packages contains these features: "gpu", did you mean: gpu?

you might want to remove the quotes from the gpu here.

Comment (jrchatruc) [reply prover/setup.sh]: That's correct, Fixed the feature flag name here [11540ff](https://github.com/matter-labs/zksync-era/pull/258/commits/11540ff3323e655ae0ce6259cb29b2903b149885)

**Review by mm-zk — APPROVED** [review]

**Review by mm-zk — APPROVED** [review]

**Pull request merged**